Main.hs

```diff
@@ -337,22 +338,33 @@ processPragma qn = getUniqueCompilerPragma pragmaName qn >>= \case
     return $ ClassPragma NoCode
   Just (CompilerPragma _ s) | s == "class" ->
     return $ ClassPragma YesCode
-  _                                        -> return DefaultPragma
-
+  Just (CompilerPragma _ s) | "deriving" `isPrefixOf` s ->
+    -- parse a deriving clause for a datatype by tacking it onto a
+    -- dummy datatype and then only keeping the deriving part
+    case Hs.parseDecl ("data X = X " ++ s) of
+      Hs.ParseFailed loc msg ->
+        setCurrentRange (srcLocToRange loc) $ genericError msg
+      Hs.ParseOk m | Hs.DataDecl _ _ _ _ _ ds <- m ->
+        return $ DerivingPragma ds
+      Hs.ParseOk _ -> __IMPOSSIBLE__
+  _ -> return DefaultPragma
 compile :: Options -> ModuleEnv -> IsMain -> Definition -> TCM CompiledDef
 compile _ _ _ def = processPragma (defName def) >>= \ case
   NoPragma -> return []
   DefaultPragma -> compile' def
   ClassPragma NoCode -> return []
   ClassPragma YesCode -> return []
+  DerivingPragma ds   -> tag <$> compileData (map (() <$) ds) def
```

Comment:
We haven't checked that the def is a `Datatype` here, so will hit the `__IMPOSSIBLE__` in `compileData` if you put `deriving` on a non-datatype. I wonder if a nicer solution would be to pass the pragma to `compile'`. That would also get rid of the duplication of `tag`.

Reply:
Thanks for spotting that! I have merged `compile` and `compile'`.